### PR TITLE
[frontend] [backend] add backend API, frontend form for editing questions

### LINKED ISSUE #10: [Backend] Set up API endpoint for editing questions
* Set up the API endpoint in the backend to handle editing (updating) of questions
  * Modify the database as needed according to the API request
  * Handle any invalid/ edge cases as necessary
 
 **Sample API Response:**
```typescript
export interface ApiResponse<D, M = {}> {
  code: null | number; // eg: 200, 400, 500...
  payload: {
		// metadata, eg: for get (index), how many entries were returned/ other relevant info
		meta: M;
		// the actual data
		data: D;
	};
  // [{type: enum- error/warning/success/info... (color of display msg), content: string}]
  messages: StatusMessage[];
}
```

### LINKED ISSUE #21: [Frontend] Edit Questions
- Enable users to edit questions. 
- Implement edit buttons. 
- Create modals/ forms to edit the question.

### PR DESCRIPTION
**Pull request by juslam19**

### CODE REVIEW

File: backend/src/question-service/routes/questions.ts [hunk 1]
```diff
@@ -13,6 +13,6 @@ router.get("/categories", QuestionsController.getQuestionCategories);
 
 router.get("/:questionId", QuestionsController.getQuestion);
 
-// TODO: "/:questionId/update"
+router.put("/:questionId/update", QuestionsController.updateQuestion);
```
Comment (sltsheryl) [inline backend/src/question-service/routes/questions.ts]: should be `router.patch("/:questionId", QuestionsController.updateQuestion);`

**Review by sltsheryl — CHANGES_REQUESTED** [review]

File: backend/src/question-service/controllers/questions.ts [hunk 2]
```diff
@@ -99,9 +99,65 @@ export const deleteQuestion: RequestHandler[] = [
...
+    // Check if a question with similar title (case insensitive) is already added.
+    const sameQuestionExists = await QuestionModel.exists({
+      title: { $regex: new RegExp("^" + formData.title + "$"), $options: "i" },
+    });
+    if (sameQuestionExists) {
+      res.status(400).json({
+        errors: [{ msg: "Question already exists in the database." }],
+      });
+      return;
+    }
```
Comment (sltsheryl) [inline backend/src/question-service/controllers/questions.ts]: for update, users should be able to update the complexity/ category/ link to question, while keeping the title the same?

currently, if the title remains the same and other fields change, it's considered invalid edit

File: backend/src/question-service/controllers/questions.ts [hunk 3]
```diff
@@ -99,9 +99,65 @@ export const deleteQuestion: RequestHandler[] = [
...
 };
 
-// TODO: updateQuestion, deleteQuestion
+// Submitting edit form 
+export const updateQuestion: RequestHandler[] = [
+  
+  param("questionId").notEmpty().withMessage("questionId field cannot be empty."),
+  param("questionId").isNumeric().withMessage("questionId should be a number."),
+  body("title").notEmpty().trim().withMessage("title cannot be empty."),
+  body("categories").isArray().withMessage("categories should be an array."),
```
Comment (seelengxd) [inline backend/src/question-service/controllers/questions.ts:122]: #28

Comment (juslam19) [reply backend/src/question-service/controllers/questions.ts:122]: Addressed in next commit, checks for invalid categories. Also added same functionality to "add question" function.

File: backend/src/question-service/controllers/questions.ts (hunk 2 above)
Comment (juslam19) [reply backend/src/question-service/controllers/questions.ts]: Addressed in next commit. Checks for duplicate title, while still allowing for same title with other changes in other fields.

File: backend/src/question-service/routes/questions.ts (hunk 1 above)
Comment (juslam19) [reply backend/src/question-service/routes/questions.ts]: Addressed in next commit.

File: backend/src/question-service/routes/questions.ts [hunk 4]
```diff
@@ -13,6 +13,6 @@ router.get("/categories", QuestionsController.getQuestionCategories);
 
 router.get("/:questionId", QuestionsController.getQuestion);
 
-router.put("/:questionId/update", QuestionsController.updateQuestion);
+router.patch("/:questionId/update", QuestionsController.updateQuestion);
```
Comment (sltsheryl) [inline backend/src/question-service/routes/questions.ts]: `router.patch("/:questionId", QuestionsController.updateQuestion);`

there shouldn't be `/update`

File: frontend/src/pages/questions/UpdateQuestion.tsx [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+              duration: 9000,
+              isClosable: true,
+            }),
+          );
+        }
+      });
+  };
+
+  return (
+    <Card m={12} p={8}>
```
Comment (seelengxd) [inline frontend/src/pages/questions/UpdateQuestion.tsx:100]: significant code duplication from create - maybe we should refactor this to its own component but if too much work maybe future pr

File: frontend/src/api/questions/questions.ts [hunk 6]
```diff
@@ -21,4 +21,19 @@ export default class QuestionsAPI {
   public async addQuestion(question: QuestionPostData): Promise<never> {
     return await client.post(this.getQuestionsUrl(), question);
   }
+
+
```
Comment (seelengxd) [inline frontend/src/api/questions/questions.ts]:
```suggestion
```
prettier is complaining locally during compilation

File: frontend/src/pages/routing/UnauthenticatedApp.tsx [hunk 7]
```diff
@@ -2,12 +2,15 @@ import React from 'react';
...
 import { CreateQuestion } from '../questions/CreateQuestion';
 import Questions from '../questions/Questions';
+import { UpdateQuestion } from '../questions/UpdateQuestion';
 
 const UnauthenticatedApp: React.FC = () => {
   return (
     <Routes>
       <Route path="questions/new" element={<CreateQuestion />} />
       <Route path="/" element={<Questions />} />
+
```
Comment (seelengxd) [inline frontend/src/pages/routing/UnauthenticatedApp.tsx]:
```suggestion
```

File: frontend/src/pages/routing/UnauthenticatedApp.tsx [hunk 8]
```diff
@@ -2,12 +2,15 @@ import React from 'react';
...
 import Questions from '../questions/Questions';
+import { UpdateQuestion } from '../questions/UpdateQuestion';
 
 const UnauthenticatedApp: React.FC = () => {
   return (
     <Routes>
       <Route path="questions/new" element={<CreateQuestion />} />
       <Route path="/" element={<Questions />} />
+      
+      <Route path="question/edit/:questionId" element={<UpdateQuestion />} />
```
Comment (seelengxd) [inline frontend/src/pages/routing/UnauthenticatedApp.tsx]:
```suggestion
      <Route path="questions/:questionId/edit" element={<UpdateQuestion />} />
```

[7 restful routes](https://gist.github.com/alexpchin/09939db6f81d654af06b)

File: frontend/src/pages/questions/UpdateQuestion.tsx [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+        setCategories(questionData.categories);
+        setComplexity(questionData.complexity);
+        setLinkToQuestion(questionData.linkToQuestion.replace(linkPrefix, ''));
+      })
+      .catch((error) => {
+        console.error('Error fetching question data:', error);
+      });
+  }, []);
+
+  const handleSubmit: React.FormEventHandler<HTMLFormElement> = (e) => {
```
Comment (seelengxd) [inline frontend/src/pages/questions/UpdateQuestion.tsx:58]: possible to upload 0 categories.
see #31

File: frontend/src/pages/questions/UpdateQuestion.tsx (hunk 5 above)
Comment (juslam19) [reply frontend/src/pages/questions/UpdateQuestion.tsx:100]: Done, refactored the code

**Comment by juslam19** [comment]
@sltsheryl Does it LGTY besides the prettier warnings?

**Comment by sltsheryl** [comment]
> @sltsheryl Does it LGTY besides the prettier warnings?

i will try to implement to pre-fill in the edit form with the previous data

**Comment by sltsheryl** [comment]
> @sltsheryl Does it LGTY besides the prettier warnings?

sorry abt the previous comment, i went to edit a non valid question so i thought there wasnt a pre filled edit form

for the backend, we are still able to access `http://localhost:3000/question/<question-id>/edit`
where `question-id` is not an existing question, should we throw error for this

**Comment by juslam19** [comment]
Yes, I intend to address that in one swoop by implementing a middleware for 404 pages. Even for non-valid pages, others are allowed to access them. This is not an isolated issue.

**Comment by sltsheryl** [comment]
currently when i go to
eg. `http://localhost:3000/question/2/edit` (a valid question)
and don't edit anything, and press submit question, i am able to edit nothing, is this the intended behavior or shd we disallow/ show a warning

**Comment by juslam19** [comment]
This behavior is harmless and does not affect the functionality of the application. While I suppose it does incur some cost in computation, I assume this is negligible.

Additionally, making the submit button not available on the frontend should be a relatively easy fix after this PR.

**Review by sltsheryl — APPROVED** [review]
previously, during the check whether duplicate question exists (for edit, i think it should include the current question itself cuz this is the case when there's no edit done) 

i think edit shd be case sensitive too cuz eg. user wants to capitalise the first letter, the edit shd be deemed as different 

otherwise, the edit form is working as intended!
(ps. as said above, shd not a form page for invalid questions - questionid that doesn't exist eg. removed questions)

**Comment by juslam19** [comment]
I believe that 'duplicate checking' actually caused some other issues, which is why I deleted it. But I will explore it, good option for implementation.

Thank you @sltsheryl .

I assume 'invalid questions' means duplicate questions?

**Comment by sltsheryl** [comment]
> I believe that 'duplicate checking' actually caused some other issues, which is why I deleted it. But I will explore it, good option for implementation.
> 
> Thank you @sltsheryl .
> 
> I assume 'invalid questions' means duplicate questions?

oh what other issues? im afraid if i overlooked if it's serious

**Comment by juslam19** [comment]
Forget I said anything. Duplicates disallowing implemented.

**Pull request merged**